netket/optimizer/sr/sr_jacobian.py

```diff
@@ -0,0 +1,223 @@
+        samples = jnp.reshape(samples, (-1, samples.shape[-1]))
+    n_samples = samples.shape[0]
+
+    if mode == "holomorphic":
+        # Preapply the model state so that when computing gradient
+        # we only get gradient of parameters
+        # Also divide through sqrt(n_samples) to normalise S matrix in the end
+        def fun(W, σ):
+            return apply_fun({"params": unravel(W), **model_state},
+                             σ[jnp.newaxis,:])[0] / n_samples**0.5
+        grads = _grad_vmap_minus_mean(fun, params, samples, True)
+    elif mode == "R2R":
+        def fun(W, σ):
+            return apply_fun({"params": unravel(W), **model_state},
+                             σ[jnp.newaxis,:])[0].real / n_samples**0.5
+        grads = _grad_vmap_minus_mean(fun, params, samples, False)
+    elif mode == "R2C":
+        def fun1(W, σ):
+            return apply_fun({"params": unravel(W), **model_state},
+                             σ[jnp.newaxis,:])[0].real / n_samples**0.5
+        def fun2(W, σ):
+            return apply_fun({"params": unravel(W), **model_state},
+                             σ[jnp.newaxis,:])[0].imag / n_samples**0.5
+        # Stack real and imaginary parts as real matrixes along the "sample"
+        # axis to get Re(O†O) directly
+        grads = jnp.concatenate(
+            (_grad_vmap_minus_mean(fun1, params, samples, False),
+             _grad_vmap_minus_mean(fun2, params, samples, False)),
+            axis = 0)
+    else:
+        raise Exception("Differentation mode must be holomorphic, R2R or R2C, got {}".format(mode))
+
+    if rescale_shift:
+        sqrt_Skk = jnp.linalg.norm(grads, axis=0, keepdims=True)
+        return grads/sqrt_Skk, 1/jnp.ravel(sqrt_Skk)
+    else:
+        return grads
+        
+@partial(jax.jit, static_argnums=(0,3))
+def _grad_vmap_minus_mean(
+        fun: Callable,
+        params: jnp.ndarray,
+        samples: jnp.ndarray,
+        holomorphic: bool
+        ):
+    """Calculates the gradient of a neural network for a number of samples 
+    efficiently using vmap(grad),
+    subtracts their mean for each parameter, i.e., each column,
+    and divides through with sqrt(#samples) to normalise the S matrix"""
+    grads = jax.vmap(jax.grad(fun, holomorphic=holomorphic), in_axes=(None,0),
+                     out_axes=0)(params, samples)
+    return grads - jnp.sum(grads, axis=0, keepdims=True) / grads.shape[0]
+
+@jax.jit
+def jacsmatrix_mat_treevec(
+    S: JacobianSMatrix, vec: Union[PyTree, jnp.ndarray]
+) -> Union[PyTree, jnp.ndarray]:
+    """
+    Perform the semi-lazy mat-vec product, where vec is either a tree with 
+    the same structure as params or a ravelled vector
+    """
+    if not hasattr(vec, "ndim"):
+        vec, unravel = nkjax.tree_ravel(vec)
+    else:
+        unravel = lambda x: x
+        
+    return unravel(
+        jnp.transpose(jnp.conj( jnp.transpose(jnp.conj(S.O @ vec)) @ S.O)) +
+        S.sr.diag_shift * vec)
```

Comment:
Before or after merging we should also add support for MPI.
If you want take up this task then have a look at how this was done in our legacy code [here](https://github.com/netket/netket/blob/4a3abfc0c88092f472a7e87ab89e6a829f5f5776/netket/legacy/optimizer/jax/stochastic_reconfiguration.py#L25) .
Each rank has its own set of samples, and so its own slice of rows of the jacobian, and essentially all that is needed is to take the mean with MPI of the result of the matrix-vector product here.
Also  the scale-inveriant regularization would have to be paralellized with MPI.

Reply:
I agree this is a very important point

Reply:
good catch

Reply:
I personally had a hard time with getting MPI to work, it always appeared to me that the different cores just do their own thing without communication. Is there a tutorial on how to actually make MPI to work?

Reply:
https://www.netket.org/docs/getting_started.html#mpi

I should write more docs on it. You can check that MPI is detected by netket by printing (nk.utils.mpi_available).
For the moment you can simply try to `pip install -U 'netket[mpi]'` and then do `mpirun -np X python Examples/Ising1d/ising.py`